### PR TITLE
Added base class for sklearn feature selection methods.

### PR DESCRIPTION
**Pull request by latture**
Adding sklearn feature selection strategies is a bit repetitive. The `fit` and `apply` logic is nearly identical between strategies. The only difference is the name of the member used to select features. In this PR I added a base class `FeatureSelectorSklearn` that defines common `fit` and `apply` methods. It relies on a `selector` that conforms to `SelectorProtocolSklearn` (a protocol also defined in this PR since sklearns selector mixin doesn't require a `fit` method) to select features. I refactored the existing sklearn feature selection methods to use the new base class. There's not really any new logic here (unless you count formalizing the methods required from a selector).

Protocols were introduced Python 3.8, so this version drops support for Python 3.7.

### CODE REVIEW

File: smlb/feature_selection/selector_protocol_sklearn.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
+from abc import abstractmethod
+from typing import Protocol, runtime_checkable
+
+import numpy as np
+from numpy import typing as npt
+
+
+@runtime_checkable
```
Comment (latture) [inline smlb/feature_selection/selector_protocol_sklearn.py:8]: This makes it possible to use `params` to check whether the selector we're given conforms to the protocol at runtime.

File: smlb/feature_selection/feature_selector_sklearn.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+    def __init__(self, selector: SelectorProtocolSklearn, *args, **kwargs):
+        """Initialize state.
+
+        Parameters:
+            selector: Feature selection method that provides ``fit`` and ``get_support`` methods.
+        """
+        super().__init__(*args, **kwargs)
+        self._selector: SelectorProtocolSklearn = params.instance(
+            selector, SelectorProtocolSklearn
```
Comment (latture) [inline smlb/feature_selection/feature_selector_sklearn.py:25]: This is much easier to check using protocols. We don't need other classes to inherit from our expected interface.

File: smlb/feature_selection/feature_selector_sklearn.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        Returns:
+            the instance itself
+        """
+        data = params.instance(data, Data)
+        n = data.num_samples
+
+        xtrain = params.real_matrix(data.samples(), nrows=n)
+        ytrain = params.real_vector(data.labels(), dimensions=n)
+
+        self._selector.fit(xtrain, ytrain)
```
Comment (latture) [inline smlb/feature_selection/feature_selector_sklearn.py:43]: before `_selector` was replaced with the sklearn member (e.g. `_select_percentile`), but the other code is the same.

File: smlb/feature_selection/feature_selector_sklearn.py (hunk 2 above)
Comment (latture) [reply smlb/feature_selection/feature_selector_sklearn.py:25]: Note, this check will fail if we don't mark the protocol as `runtime_checkable`.

**Review by maxhutch — APPROVED** [review]
Neat!

**Pull request merged**